packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts

```diff
@@ -0,0 +1,69 @@
+type Env = 'edge' | 'stable';
+
+type SiteType = 'simple' | 'atomic';
+
+type Variation = 'siteEditor';
+
+type Feature = 'gutenberg' | 'coblocks';
+type FeatureKey = { [ key in Feature ]?: Env | undefined } & {
+	siteType: SiteType;
+	variation?: Variation;
+};
+export type FeatureCriteria = FeatureKey & { accountName: string };
+type FeatureMap = Map< string, string >;
+
+function stringifyKey( o: FeatureKey ) {
+	const keys = Object.keys( o ) as [ keyof FeatureKey ];
+	const sorted = keys.sort().reduce( ( sorted, key ) => {
+		sorted[ key ] = o[ key ];
+		return sorted;
+	}, {} as any ) as FeatureKey;
+
+	return JSON.stringify( sorted );
+}
+
+function criteriaToMap( criteria: FeatureCriteria[], map: FeatureMap ): FeatureMap {
+	return criteria.reduce( ( acc, criteria ) => {
+		const { accountName, ...rest } = criteria;
+		acc.set( stringifyKey( rest ), accountName );
+		return acc;
+	}, map );
+}
```

Comment:
This whole implementation is really slick! 💪  This definitely seems like it could be extended really easily!

Is this potentially overkill though for our majority use cases? I'm definitely not saying that's the case, but wanted to open up the discussion. 

The way I see it, the vast majority of scripts are going to do (or could be adjusted to do) one of three things:
1. Just use the default user (this is Calypso-unique flows)
2. Use the same shared set of logic to select a Gutenberg user based on values from `envVariables`
3. Be unique enough that it has its own custom set of users and unique logic to pick them

If that's the case, would it be easier to just make a function that handles the no. 2 case by taking in an `envVariables` object and returning a user?

I wonder if that might simplify things for most test writers because they don't need to repeat all the ternary expressions to create the `FeatureKey` object, or have to worry about making sure there's an exact match of Feature values in the `defaultCriteria` table.

I see two potential benefits of this:
1. A lot of test scripts could just start with a call that looks like...
```typescript
const testAccount = getGutenbergTestAccount( envVariables );
``` 
2. It might make the logical flow a little bit easier to follow. I realize this is likely very much my personal preference, but I find it a little easier to follow a logical flow in statements than matching up all the keys and values in a table. 

Again, I'm not entirely sure I even agree with myself here. 😆  And this current approach is really cool! I just wanted to double check there wasn't an easier option out there! I hope that's not unhelpful!

Reply:
As a middle ground to this too, we could also just add a second function that transforms an `envVariables` object to a `FeatureKey` object, so that we would have to replicate the ternary logic as much in all our Gutenberg scripts.

Reply:
> Is this potentially overkill though for our majority use cases? I'm definitely not saying that's the case, but wanted to open up the discussion.

Yeah, that crossed my mind! I think we could definitely get away without having such an abstraction, especially for the "Be unique enough that it has its own custom set of users and unique logic to pick them" scenario. Right now - before we add the AT test "dimension" - it could indeed be considered overkill. As soon as I started asking the "Is this an AT test" question for each test, things started getting a bit messy, though:

```
let accountName: string;
if ( envVariables.TEST_ON_ATOMIC ) {
	accountName = 'testonatomicuser';
} else if ( envVariables.GUTENBERG_EDGE ) {
	accountName = 'gutenbergSimpleSiteEdgeUser';
} else {
	accountName = 'gutenbergSimpleSiteUser';
}
```

And soon we'll probably also extend CoBlock's tests to run on AT, and this would become a bit more complex:

```
let accountName: string;
if ( envVariables.TEST_ON_ATOMIC ) {
 accountName = 'regulartestonatomicuser';
} else {
 if ( envVariables.COBLOCKS_EDGE ) {
 	accountName = 'coBlocksSimpleSiteEdgeUser';
 } else if ( envVariables.GUTENBERG_EDGE ) {
	 accountName = 'gutenbergSimpleSiteEdgeUser';
 } else {
	accountName = 'gutenbergSimpleSiteUser';
 }
}
```

It's possible that we might extend this concept to test a future version of another plugin in AT (less likely, but who knows). In fact, we'll probably have two sites for GB and AT (right now we only have one, for edge) and will allow the `GUTENBERG_EDGE` flag to influence which AT site is selected, so the logic with conditionals would look something like:

```
let accountName: string;
if ( envVariables.TEST_ON_ATOMIC ) {
	accountName = 'testonatomicuser';
} else if ( envVariables.GUTENBERG_EDGE ) {
	accountName = 'gutenbergSimpleSiteEdgeUser';
} else {
	accountName = 'gutenbergSimpleSiteUser';
}
```

And soon we'll probably also extend CoBlock's tests to run on AT, and this would become a bit more complex:

```
let accountName: string;
if ( envVariables.TEST_ON_ATOMIC ) {
  if ( envVariables.GUTENBERG_EDGE ) {
    accountName = 'atgbedgeaccount';
  } else {
   accountName = 'atstablegbaccount';
  }
} else {
 if ( envVariables.COBLOCKS_EDGE ) {
 	accountName = 'coBlocksSimpleSiteEdgeUser';
 } else if ( envVariables.GUTENBERG_EDGE ) {
	 accountName = 'gutenbergSimpleSiteEdgeUser';
 } else {
	accountName = 'gutenbergSimpleSiteUser';
 }
}
```

It could possibly be tidied up with ternary expressions and some helpers locally maybe but at the expense of not being DRY. It's not a huuuge deal, but is a bit painful to be doing this everytime and to read such a long chain of conditional statements. That's why I thought a declarative way - something like a table - would be better - it just feels more natural to reason about and write. The "syntax" of each criterion is basically the following question:

* Does the state of the world look like x? Then account = y. `x` being `FeatureKey` and `y` being the `accoutName`. Ex: `{ gutenberg: 'edge', coblocks: 'edge', siteType: 'simple', accountName: 'accountForGBEdgeAndCoBlocksEdge' }`.

This would have the benefit of being easy to extend for any other scenarios where you need a site to be conditional on external flags.

That fits the pattern of a test site per configuration we follow now very naturally. We are basically writing conditions for when this site is selected in a table of objects instead of using conditionals, which in my opinion is easier to read and scan through, with the nice side effect of DRYing up the client code to select the account.

> but I find it a little easier to follow a logical flow in statements than matching up all the keys and values in a table.

For most simple cases where the outcome is based on a single flag (or might not need a flag at all) then I agree. The more variables in the equation, the messier the code will become, though. It's possible we might not go beyond adding the `TEST_ON_ATOMIC` variable, but what if have more specific test sites in the future? For example, what if we add a new parameter that passes a theme (for the most important, most used themes) to the test and we then have specific sites for:
* theme x + site editor + gutenberg edge on simple
* theme x + site editor + gutenberg stable on atomic

This would get out of control with `if` conditionals.

Reply:
> const testAccount = getGutenbergTestAccount( envVariables );

I think you make a very good point here! I'm tempted to just take this path, as 99% of the time the data will come from the env variables.

Reply:
Another benefit of using such a function is to create a common interface to parameterize tests based on the env variables. I.e using the theme example above. Let's say we want to start running test x in theme y when gutenberg is edge and stable. We can just add a new "theme" string parameter to TC and to the type behind `envVariables`. If we change the function to just accept the env variables object, we'd just need to add new entries to the table (and update the `FeatureKey` type to have a new `theme` attribute of type `string`):

* `{ gutenberg: 'edge', siteType: 'simple', theme: 'themex', accountName: 'edgeThemeXSimple'}`
* `{ gutenberg: 'stable', siteType: 'simple', theme: 'themex', accountName: 'edgeThemeXSimple'}`

Reply:
And I'll leave it up to you whether having middleware that auto-makes the ENV --> `FeatureSet` conversion helps or hurts the readability/maintainability!

Reply:
I think it's worth adding that helper function, but I'll work on it in a follow-up.